### ISSUE TITLE
GPU crash at start of masked (focused) refinment 

### ISSUE THREAD

**Issue by OmidGHCryoEM** [3dem/relion#668]
Dear Relion users,

I recently encountered the following problem:
Refine3D run seem to crash after I try to run a masked 3D refinement . This crash only occurred after I updated my Relion to 3.1. The previous version ran just fine. 

**Environment:**
 - OS: Centos
 - MPI runtime: OpenMPI 2.0.2
 - RELION version: 3.1.0-commit-1349c5
 - Memory: 376 GB
 - GPU: 4X RTX 2080

**Dataset:**
 - Box size: 448
 - Pixel size: 1.09
 - Number of particles: 289,000
 - Description: A tetrameric protein of about 2.2MDa in total

**Job options:**
 - Type of job: Refine3D masked
 - Number of MPI processes: 5
 - Number of threads: 12
 - Full command (see `note.txt` in the job directory): `which relion_refine_mpi` --continue Refine3D/job204/run_ct3_it010_optimiser.star --o Refine3D/job223/run_ct10 --dont_combine_weights_via_disc --scratch_dir /scr/ --pool 6 --pad 2  --particle_diameter 420 --solvent_mask MaskCreate/job185/mask.mrc --solvent_correct_fsc  --j 12 --gpu ""  --pipeline_control Refine3D/job223/

**Error message:**
ERROR: out of memory in //usr/local/relion-3-source/src/acc/cuda/custom_allocator.cuh at line 436 (error-code 2)
in: //usr/local/relion-3-source/src/acc/cuda/cuda_settings.h, line 67
ERROR: 

A GPU-function failed to execute.

 If this occured at the start of a run, you might have GPUs which
are incompatible with either the data or your installation of relion.
If you 

	-> INSTALLED RELION YOURSELF: if you e.g. specified -DCUDA_ARCH=50
	   and are trying ot run on a compute 3.5 GPU (-DCUDA_ARCH=3.5), 
	   this may happen.

	-> HAVE MULTIPLE GPUS OF DIFFERNT VERSIONS: relion needs GPUS with
	   at least compute 3.5. You may be trying to use a GPU older than
	   this. If you have multiple generations, try specifying --gpu <X>
	   with X=0. Then try X=1 in a new run, and so on. The numbering of
	   GPUs may not be obvious from the driver or intuition. For a list
	   of GPU compute generations, see 

	   en.wikipedia.org/wiki/CUDA#Version_features_and_specifications

	-> ARE USING DOUBLE-PRECISION GPU CODE: relion was been written so
	   as to not require this, and may thus have unforeseen requirements
	   when run in this mode. If you think it is nonetheless necessary,
	   please consult the developers with this error.

If this occurred at the middle or end of a run, it might be that

	-> YOUR DATA OR PARAMETERS WERE UNEXPECTED: execution on GPUs is 
	   subject to many restrictions, and relion is written to work within
	   common restraints. If you have exotic data or settings, unexpected
	   configurations may occur. See also above point regarding 
	   double precision.
If none of the above applies, please report the error to the relion
developers at    github.com/3dem/relion/issues

=== Backtrace  ===
/usr/local/relion-3/bin/relion_refine_mpi(_ZN11RelionErrorC1ERKSsS1_l+0x41) [0x470ae1]
/usr/local/relion-3/bin/relion_refine_mpi() [0x60db89]
/usr/local/relion-3/bin/relion_refine_mpi(_ZN9Projector26computeFourierTransformMapER13MultidimArrayIdES2_iibbiPKS1_b+0x2b46) [0x611f86]
/usr/local/relion-3/bin/relion_refine_mpi(_ZN7MlModel23setFourierTransformMapsEbidPK13MultidimArrayIdE+0x902) [0x4fa152]
/usr/local/relion-3/bin/relion_refine_mpi(_ZN11MlOptimiser16expectationSetupEv+0x5e) [0x59ca4e]
/usr/local/relion-3/bin/relion_refine_mpi(_ZN14MlOptimiserMpi11expectationEv+0x551) [0x44b811]
/usr/local/relion-3/bin/relion_refine_mpi(_ZN14MlOptimiserMpi7iterateEv+0xc1) [0x45b021]
/usr/local/relion-3/bin/relion_refine_mpi(main+0x5f) [0x43281f]
/lib64/libc.so.6(__libc_start_main+0xf5) [0x7f071e1da555]
/usr/local/relion-3/bin/relion_refine_mpi() [0x4364ef]
==================
ERROR: 

A GPU-function failed to execute.

 If this occured at the start of a run, you might have GPUs which
are incompatible with either the data or your installation of relion.
If you 

	-> INSTALLED RELION YOURSELF: if you e.g. specified -DCUDA_ARCH=50
	   and are trying ot run on a compute 3.5 GPU (-DCUDA_ARCH=3.5), 
	   this may happen.

	-> HAVE MULTIPLE GPUS OF DIFFERNT VERSIONS: relion needs GPUS with
	   at least compute 3.5. You may be trying to use a GPU older than
	   this. If you have multiple generations, try specifying --gpu <X>
	   with X=0. Then try X=1 in a new run, and so on. The numbering of
	   GPUs may not be obvious from the driver or intuition. For a list
	   of GPU compute generations, see 

	   en.wikipedia.org/wiki/CUDA#Version_features_and_specifications

	-> ARE USING DOUBLE-PRECISION GPU CODE: relion was been written so
	   as to not require this, and may thus have unforeseen requirements
	   when run in this mode. If you think it is nonetheless necessary,
	   please consult the developers with this error.

If this occurred at the middle or end of a run, it might be that

	-> YOUR DATA OR PARAMETERS WERE UNEXPECTED: execution on GPUs is 
	   subject to many restrictions, and relion is written to work within
	   common restraints. If you have exotic data or settings, unexpected
	   configurations may occur. See also above point regarding 
	   double precision.
If none of the above applies, please report the error to the relion
developers at    github.com/3dem/relion/issues

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 4

**Comment by biochem-fan** [comment]
Can you try the `devel-nvidia-optim4` branch? This should solve the issue. We are testing this branch before merging to ver3.1 branch.

**Comment by OmidGHCryoEM** [comment]
Thank you for the prompt reply. Can you clarify what you mean. I am not sure what devel-nvidia-optim4 branch is. Is that another build for Relion? In which case I should just update my current build?

**Comment by biochem-fan** [comment]
Please do this:

```
git fetch -a
git checkout devel-nvidia-optim4
git pull
cmake ..
make
```

**Comment by OmidGHCryoEM** [comment]
Thanks for the clarification. Tried it again after the new install. Same error. Is there a way to confirm the branch build? The header still says 3.1.0-commit-1349c5. Unless I did something wrong?

**Comment by biochem-fan** [comment]
You are still using the old version. What does `git branch` say?

**Comment by OmidGHCryoEM** [comment]
* devel-nvidia-optim4
  master
  ver3.1

**Comment by biochem-fan** [comment]
Hmm, this is strange.

It should be ecf51e5. https://github.com/3dem/relion/commits/devel-nvidia-optim4

Did you make sure your PATH points to the right version?

**Comment by OmidGHCryoEM** [comment]
So I installed the binaries in /usr/local/relion-3 

by using the following
cmake -DCUDA_ARCH=75 -DFORCE_OWN_FLTK=ON -DCMAKE_INSTALL_PREFIX=/usr/local/relion-3
make -j8
This is just how I installed relion previously.  But for some reason when I try it now the binaries are installed in the source folder. 
So naturally when I try to run relion it is still running the old binaries and not the new ones. 

Sorry for the confusion. I copied over the correct binaries and library files and restarted refinement now.
